src/saml.ts

```diff
@@ -1077,37 +1077,45 @@ class SAML {
           }
         }
 
-        subjectConfirmation = subject[0].SubjectConfirmation
-          ? subject[0].SubjectConfirmation[0]
-          : null;
-        confirmData =
-          subjectConfirmation && subjectConfirmation.SubjectConfirmationData
-            ? subjectConfirmation.SubjectConfirmationData[0]
-            : null;
-        if (subject[0].SubjectConfirmation && subject[0].SubjectConfirmation.length > 1) {
-          msg = "Unable to process multiple SubjectConfirmations in SAML assertion";
-          throw new Error(msg);
-        }
-
-        if (subjectConfirmation) {
-          if (confirmData && confirmData.$) {
-            const subjectNotBefore = confirmData.$.NotBefore;
-            const subjectNotOnOrAfter = confirmData.$.NotOnOrAfter;
-            const maxTimeLimitMs = this.processMaxAgeAssertionTime(
-              this.options.maxAssertionAgeMs,
-              subjectNotOnOrAfter,
-              assertion.$.IssueInstant
-            );
-
-            const subjErr = this.checkTimestampsValidityError(
-              nowMs,
-              subjectNotBefore,
-              subjectNotOnOrAfter,
-              maxTimeLimitMs
-            );
-            if (subjErr) {
-              throw subjErr;
+        if (subject[0].SubjectConfirmation && subject[0].SubjectConfirmation?.length > 0) {
+          subjectConfirmation = subject[0].SubjectConfirmation?.find(
+            (subjectConfirmation: XMLOutput) => {
+              let res = false;
+              if (subjectConfirmation) {
+                confirmData =
+                  subjectConfirmation && subjectConfirmation.SubjectConfirmationData
+                    ? subjectConfirmation.SubjectConfirmationData[0]
+                    : null;
+                if (confirmData && confirmData.$) {
+                  const subjectNotBefore = confirmData.$.NotBefore;
+                  const subjectNotOnOrAfter = confirmData.$.NotOnOrAfter;
+                  const maxTimeLimitMs = this.processMaxAgeAssertionTime(
+                    this.options.maxAssertionAgeMs,
+                    subjectNotOnOrAfter,
+                    assertion.$.IssueInstant
+                  );
+
+                  const subjErr = this.checkTimestampsValidityError(
+                    nowMs,
+                    subjectNotBefore,
+                    subjectNotOnOrAfter,
+                    maxTimeLimitMs
+                  );
+                  if (subjErr === null) res = true;
+                }
+              }
+              return res;
             }
+          );
+
+          if (subjectConfirmation) {
+            subjectConfirmation = subjectConfirmation[0];
+            confirmData =
+              subjectConfirmation && subjectConfirmation.SubjectConfirmationData
+                ? subjectConfirmation.SubjectConfirmationData[0]
+                : null;
+          } else {
```

Comment:
The tests pass without this `else` block, so either we're missing a test, or something is wrong.

It seems like something might be wrong. In the old code, if `subject[0].SubjectConfirmation` can't be used `subjectConfirmation` is set equal to `null`. In this code, it isn't. This `else` block should probably be set to match line 1080, not line 1111.

It actually seems the course of wisdom to initialize `subjectConfirmation` and `confirmData` before the check for length, just as before, so that if length turns out to be 1, the old behavior is preserved. If the length is >1, we fall in to this new code. This would require that `subjectConfirmation` not be used with your `.find()` to hold the intermediate results.

In this way, you only have to add tests for new new condition you're affecting, which you've done. With the changes you've made here, I'd like to see some more tests assuring the existing behavior.

Reply:
You're right; Probably I was drunk to write that piece of code. The working code should read like: "if there is any subjectconfirmation then filter them for valid timestamp and give back the resulting array for further validation". So the "else" part is really a useless piece of code. I really apologize for this.